gapir/replay_service/service.proto

```diff
@@ -127,7 +134,8 @@ message Notification {
   uint64 id = 1;
   oneof val {
     ErrorMessage error_msg = 2;
-    Data data = 3;
+    ReplayStatus replay_status = 3;
```

Comment:
I would prefer we add another ReplayResponse, instead of overloading Notification more. Having special values for the ID of notifications will get confusing and harder to maintain.

Reply:
I see, make sense! Will try to refactor the code to address this problem.

Reply:
Hey Andrew, as I really put my hands on this work, trying to make it a separate ReplayResponse, I found I may include some more code complexity in this way. Some of the observations and thoughts are given below. Would you mind help me see whether these make sense and shall I keep the structure this way?

1.There's already many handlers to deal with in our builder. 
I will have to create another handler and return it in builder.Build(), which will make our code heavy and complex.
2.The definition of Notification: "sending data from GAPIR to GAPIS" match my needs here exactly. 
It simply handle data transfer and I don't see ways to make it more concise. I feel I will write a new set of exactly the same code if I really do it.
3.The notification framework is written very well but not used much.
Currently the customized sendNotificationData() is not called in our codebase (from my observation), the only notification method in use is sendErrorMsg(). I feel it is a waste not putting it into use and generate another logic line.

Reply:
So the notification is meant to be free-form data. Basically a bunch of Bytes. (Similar to how Post works) we extended it to work with error messages, I am just worried about adding more types of structured data to it.

That being said we could update it with a slightly different notification setup. I just dont like the idea of overloading ID with more special values. WE could move ID into the "Data" message, and leave the other notifications out. That would be fine with me.

Reply:
For example:

```
// Data is a message contains data that a GAPIR device collected during the
// replay.
message Data {
  uint64 id = 1;
  uint64 label = 2;
  bytes data = 3;
}

// Notification is a message that GAPIR device wants to send to GAPIS. This
// message can be used to send back data such as error message or other data to
// GAPIS.
message Notification {
  oneof val {
    ErrorMessage error_msg = 1;
    ReplayStatus replay_status = 2;
    Data data = 3;
  }
}
```

That way we can switch on `val.(type)` rather than an arbitrary ID that means a special thing.

Reply:
The original modification to add an `ID` to `Notification` is to map the `ID` to the `notificaitonReader`, if we switch on type, then we need to find a  way to map the notification with its reader.

Reply:
I see... So if we move the ID into Data, and switch on val.(type):
The way we carry Notification Readers in Builder might transfer to 
`	notificationReaderReplayStatus NotificationReader
	notificationReaderErrorMessage NotificationReader
	notificationReadersData map[uint64]NotificationReader`
to still be able to pinpoint the reader without an ID.

It can transfer the complexity from constants to builder structure.
May I ask which practice will benefit more in the future?
And how do we plan to use the customized Notification.Data? (If we don't use it we can directly get away with the map)

Reply:
Sounds reasonable. And maybe we can generate a new struct NotificationReaders to wrap around all the readers to organize things better. 

If Robin also agrees with it, I can do this according to our discussion. ^_^